### ISSUE TITLE
Using @JsonSubTypes fails on jackson serialization/deserialization

### ISSUE THREAD

**Issue by pnepywoda** [immutables/immutables#175]
Create the following 2 classes:

```java
@JsonSubTypes({
        @Type(DatasetIdLocator.class),
        @Type(DatasetPathLocator.class),
        @Type(ImmutableDatasetIdLocator.class),
        @Type(ImmutableDatasetPathLocator.class)
    })
@JsonTypeInfo(use = Id.NAME, include = As.PROPERTY, property = "type", visible = false)
public interface DatasetLocator {

}
```

```java
@Value.Immutable
@Value.Style(visibility = Value.Style.ImplementationVisibility.PACKAGE)
@JsonSerialize(as = ImmutableDatasetIdLocator.class)
@JsonDeserialize(as = ImmutableDatasetIdLocator.class)
public interface DatasetIdLocator extends DatasetLocator {

    String getDatasetId();

    class Builder extends ImmutableDatasetIdLocator.Builder {}

}
```

And test with:
```java
    @Test
    public void testJsonSerDe() throws IOException {
        DatasetIdLocator locator = new DatasetIdLocator.Builder().datasetId("id").build();
        ObjectMapper om = new ObjectMapper();
        DatasetIdLocator value = om.readValue(om.writeValueAsBytes(locator), DatasetIdLocator.class);
        assertThat(locator.getDatasetId(), is(value.getDatasetId()));
    }
```

The resulting exception is:
```
com.fasterxml.jackson.databind.JsonMappingException: Could not resolve type id 'ImmutableDatasetIdLocator$Json' into a subtype of [simple type, class com.palantir.foundry.build.api.DatasetIdLocator]: known type ids = [DatasetIdLocator, DatasetPathLocator, ImmutableDatasetIdLocator, ImmutableDatasetPathLocator]
 at [Source: [B@4d0d9fe7; line: 1, column: 2]
	at com.fasterxml.jackson.databind.JsonMappingException.from(JsonMappingException.java:148)
	at com.fasterxml.jackson.databind.DeserializationContext.unknownTypeException(DeserializationContext.java:964)
	at com.fasterxml.jackson.databind.jsontype.impl.TypeDeserializerBase._handleUnknownTypeId(TypeDeserializerBase.java:277)
	at com.fasterxml.jackson.databind.jsontype.impl.TypeDeserializerBase._findDeserializer(TypeDeserializerBase.java:159)
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer._deserializeTypedForId(AsPropertyTypeDeserializer.java:108)
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer.deserializeTypedFromObject(AsPropertyTypeDeserializer.java:93)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeWithType(BeanDeserializerBase.java:992)
	at com.fasterxml.jackson.databind.deser.impl.TypeWrappedDeserializer.deserialize(TypeWrappedDeserializer.java:42)
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:3702)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2807)
```

This happened sometime between 2.0.17 and 2.0.21 (was working as of 2.0.17). Looks like it was part of https://github.com/immutables/immutables/commit/ce76e4f29bd134b8afe09290b9313a512c3108e3

**Comment by elucash** [comment]
Thank you for the bug report! It could be a Jackson bug or rather undefined behavior of combination of features, but we will also try to mitigate it on Immutables side. I've reposted it as FasterXML/jackson-databind/issues/937 to clarify behavior in this case.

**Comment by elucash** [comment]
So while it's Jackson shortcoming, it is hardly a bug, but expected behavior. We'll try to work around some changes made is ce76e4f29bd134b8afe09290b9313a512c3108e3 
Also #167 is affected by this

**Comment by pnepywoda** [comment]
Will this issue be fixed in the immutables project then? It looks like the jackson-databind issue was closed.

**Comment by elucash** [comment]
yes, we will try to fix this on the side of immutables, just not clear how exactly — if we revert back to the previous serialization strategy, we have to make sure to fix original issues with it.

**Comment by elucash** [comment]
Fix Released in 2.1.0
